### PR TITLE
Add product comparison feature

### PR DESCRIPTION
**Pull request by skiliton**

### CODE REVIEW

**Review by JuliaNortman — APPROVED** [review]

File: marketplace-client/src/marketplace/src/app/_components/alert/alert.component.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  selector: 'app-alert',
+  templateUrl: './alert.component.html',
+  styleUrls: ['./alert.component.css']
+})
+export class AlertComponent{
+
+  alerts: Alert[] = [];
+
+  constructor(private alertService: AlertService) {
+    this.alerts = alertService.getAlerts();
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/_components/alert/alert.component.ts:15]: https://indepth.dev/posts/1119/the-essential-difference-between-constructor-and-ngoninit-in-angular

File: marketplace-client/src/marketplace/src/app/_components/alert/alert.component.html [hunk 2]
```diff
@@ -1,3 +1,8 @@
-<div class="alert alert-primary" role="alert">
-
+<div class="alert-container">
+  <div *ngFor="let alert of alerts" class="alert alert-{{alert.type}} alert-dismissible fade show" role="alert">
+    {{alert.message}}
+    <button (click)="close(alert)" type="button" class="close" data-dismiss="alert" aria-label="Close">
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/_components/alert/alert.component.html:4]: good practice is 1 attr = 1 line for readability

File: marketplace-client/src/marketplace/src/app/app.component.html [hunk 3]
```diff
@@ -35,9 +35,9 @@
       <a>DELIVERY</a>
     </div>
     <div class="header-right">
-      <div class="compare-icon">
+      <div routerLink="products/comparison" class="compare-icon">
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/app.component.html:38]: what if `comparedProd.length === 0` and it's nothing to compare? do we need to show this link at all?

File: marketplace-client/src/marketplace/src/app/app.component.ts [hunk 4]
```diff
@@ -12,7 +15,13 @@ export class AppComponent {
...
 
-  constructor(private accountService: AccountService) {
+  comparedProd: Product[] = [];
+
+  constructor(
+    private accountService: AccountService,
+    @Inject(LimitedProductComparisonService)private comparisonService: ProductComparisonService
+  ) {
+    this.comparedProd = comparisonService.getProducts();
     this.accountService.account.subscribe(x => this.user = x);
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/app.component.ts:25]: - don't forget to unsubscribe
- don't do complex logic in a constructor

File: marketplace-client/src/marketplace/src/app/app.component.ts [hunk 5]
```diff
@@ -12,7 +15,13 @@ export class AppComponent {
   // @ts-ignore
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/app.component.ts:15]: why do we need this `@ts-ignore`?

File: marketplace-client/src/marketplace/src/app/app.module.ts [hunk 6]
```diff
@@ -38,7 +39,7 @@ import { AlertComponent } from './alert/alert.component';
     FormsModule,
     ToastrModule.forRoot(),
   ],
-  declarations: [AppComponent, HomeComponent, CartComponent, AlertComponent],
+  declarations: [AppComponent, HomeComponent, CartComponent, AlertComponent, ProductComparisonComponent],
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/app.module.ts:42]: 1 declaration = 1 line is a good practice

File: marketplace-client/src/marketplace/src/app/product-catalog/product/product.component.html [hunk 7]
```diff
@@ -15,7 +15,10 @@ <h6 class="card-subtitle mb-2 text-muted">Firm: {{product.firmName}}</h6>
             <div class="price text-success">
               <h5 class="mt-4">{{product.price}}$/{{product.unit}}</h5>
             </div>
-            <a (click)="addToCart()" class="btn btn-danger mt-3"><i class="fas fa-shopping-cart"></i> Add to Cart</a>
+            <div>
+              <button type="button" (click)="addToCart()" class="btn btn-danger mt-3"><i class="fas fa-shopping-cart"></i> Add to Cart</button>
+              <button type="button" (click)="addToComparison()" class="btn btn-info mt-3"><i class="fas fa-balance-scale"></i> Add to Comparison</button>
+            </div>
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/product-catalog/product/product.component.html:21]: new tag = new line + proper line indent. It shows folded structure and hierarchy

File: marketplace-client/src/marketplace/src/app/_services/product-comparison/page-product-comparison.service.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    if(this.products.indexOf(product) == -1){
+      this.products.push(product);
+    }
+  }
+
+  setProducts(products: Product[]){
+    this.products.splice(0);
+    products.forEach((product)=>{
+      this.addProduct(product);
+    },this)
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/_services/product-comparison/page-product-comparison.service.ts:22]: why's logic so complex? think about the number of operations performed...
Just copy 1 array in another, it's 1 line, 1 operation

https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Operators/Spread_syntax

File: marketplace-client/src/marketplace/src/app/_services/product-comparison/limited-product-comparison.service.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import {Inject, Injectable} from '@angular/core';
+import {ProductComparisonService} from "./product-comparison";
+import {PageProductComparisonService} from "./page-product-comparison.service";
+import {Product} from "../../_models/products/product";
+import {AlertService} from "../alert.service";
+import {AlertType} from "../../_models/alert";
+import {BrowserProductComparisonService} from "./browser-product-comparison.service";
+
+const maxProducts = 4;
```
Comment (MaxxMorozov) [inline marketplace-client/src/marketplace/src/app/_services/product-comparison/limited-product-comparison.service.ts:9]: why is it not a member of the class?

**Pull request merged**